blockexplorer/store/migrator.go

```diff
@@ -0,0 +1,245 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package store
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/georgysavva/scany/pgxscan"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+// Migrator is a data migration agent that will migrate tx_results data
+// that is required for providing data for the block explorer APIs.
+// The tx_results table is populated by Tendermint and the number of records
+// can be very large. This causes data migration difficult when there is a need
+// to add additional information to tx_results in order to fulfill requirements
+// for the block explorer APIs.
+// The migration agent will allow migrations to run in the background without
+// blocking upgrades with slow database migration scripts.
+type Migrator struct {
+	pool        *pgxpool.Pool
+	migrateData bool
+}
+
+// NewMigrator creates a new data migration agent.
+func NewMigrator(pool *pgxpool.Pool, migrateData bool) *Migrator {
+	return &Migrator{
+		pool:        pool,
+		migrateData: migrateData,
+	}
+}
+
+func (m *Migrator) checkCanMigrate() bool {
+	// We only want to migrate if we have a tx_results_old table
+	sql := `select table_name from information_schema.tables where table_name = 'tx_results_old'`
+	var tableName string
+	if err := m.pool.QueryRow(context.Background(), sql).Scan(&tableName); err != nil {
+		return false
+	}
+	return true
+}
+
+func (m *Migrator) cleanupOldData() error {
+	// we want to drop the old table if it exists
+	sql := `drop table if exists tx_results_old`
+	if _, err := m.pool.Exec(context.Background(), sql); err != nil {
+		return fmt.Errorf("could not drop old table: %w", err)
+	}
+
+	return nil
+}
+
+// Migrate will run the data migration.
+func (m *Migrator) Migrate() error {
+	if !m.checkCanMigrate() {
+		return nil
+	}
+
+	// create indexes on the tables that we will be querying
+	if err := m.createIndexes(); err != nil {
+		return err
+	}
+
+	// get a list of dates that we need to migrate
+	migrateDates, err := m.getMigrationDates()
+	if err != nil {
+		return err
+	}
+
+	// loop through each date and migrate the data for the date
+	for _, d := range migrateDates {
+		// we're going to make use of temporary tables which are only visible to the session that created them
+		// we therefore have to use a single connection and make sure that the same connection is used for the
+		// migration process.
+		conn, err := m.pool.Acquire(context.Background())
```

Comment:
Actually, I'm not sure about the use of the background index here, shouldn't we pass in/use the same context with which we call `Migrate` in `blockexplorer.go`?